### ISSUE TITLE
Postgres db migration fails (260.6 -> 261) 

### ISSUE THREAD

**Issue by cnelson** [cloudfoundry/bosh#1591]
Updating a bosh director caused it to not come back up with`PG::Error: ERROR:  index "blobstore_id_sha1_idx" does not exist (Sequel::DatabaseError)
` error in `migrate.stderr.log`.  Full stack trace below.

    /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/adapters/postgres.rb:186:in `exec': PG::Error: ERROR:  index "blobstore_id_sha1_idx" does not exist (Sequel::DatabaseError)
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/adapters/postgres.rb:186:in `block in execute_query'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/logging.rb:35:in `log_yield'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/adapters/postgres.rb:186:in `execute_query'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/adapters/postgres.rb:173:in `block in execute'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/adapters/postgres.rb:149:in `check_disconnect_errors'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/adapters/postgres.rb:173:in `execute'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/adapters/postgres.rb:525:in `_execute'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/adapters/postgres.rb:341:in `block (2 levels) in execute'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/adapters/postgres.rb:546:in `check_database_errors'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/adapters/postgres.rb:341:in `block in execute'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/connecting.rb:251:in `block in synchronize'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/connection_pool/threaded.rb:101:in `hold'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/connecting.rb:251:in `synchronize'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/adapters/postgres.rb:341:in `execute'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/query.rb:52:in `execute_dui'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/query.rb:45:in `execute_ddl'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/schema_methods.rb:420:in `block in apply_alter_table'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/schema_methods.rb:420:in `each'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/schema_methods.rb:420:in `apply_alter_table'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/schema_methods.rb:425:in `apply_alter_table_generator'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/schema_methods.rb:81:in `alter_table'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/bosh-director-261.0.0/db/migrations/director/20170119202003_update_sha1_column_sizes.rb:37:in `block (2 levels) in <top (required)>'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/extensions/migration.rb:106:in `instance_eval'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/extensions/migration.rb:106:in `apply'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/extensions/migration.rb:668:in `block (2 levels) in run'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/transactions.rb:136:in `_transaction'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/transactions.rb:110:in `block in transaction'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/connecting.rb:251:in `block in synchronize'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/connection_pool/threaded.rb:105:in `hold'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/connecting.rb:251:in `synchronize'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/database/transactions.rb:99:in `transaction'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/extensions/migration.rb:470:in `checked_transaction'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/extensions/migration.rb:667:in `block in run'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/extensions/migration.rb:664:in `each'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/sequel-4.32.0/lib/sequel/extensions/migration.rb:664:in `run'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/bosh-director-261.0.0/bin/bosh-director-migrate:36:in `migrate'
    from /var/vcap/packages/director/gem_home/ruby/2.3.0/gems/bosh-director-261.0.0/bin/bosh-director-migrate:41:in `<top (required)>'
    from /var/vcap/packages/director/bin/bosh-director-migrate:16:in `load'
    from /var/vcap/packages/director/bin/bosh-director-migrate:16:in `<main>'

**Comment by tylerschultz** [comment]
https://www.pivotaltracker.com/story/show/140068815

**Comment by tylerschultz** [comment]
@cnelson Is it possible for you list the indexes you have on the `local_dns_blobs` table?

Something like this:
```
# ssh into your director vm and become root

$ /var/vcap/packages/postgres/bin/psql -U postgres

postgres=> \c bosh
postgres=> select * from pg_indexes where tablename = 'local_dns_blobs';
```

**Comment by cnelson** [comment]
@tylerschultz Our db was missing the index, adding it allowed the migration to continue.

I believe this issued happened because `v257` shipped this version of the db migration: https://github.com/cloudfoundry/bosh/commit/85bb2ecfce08bdc022be31ed8dbfcbfd02568164#diff-c285b387ef4e9a9c896a4ce71d990073 and then `v258` shipped this version (which added this index): https://github.com/cloudfoundry/bosh/commit/35587c4d58de9998f0e93b79b97860a7cbbc7a96 instead of adding a new migration file to create it.

Sequel tracks which migrations it's applied and will not apply them again, so the updated migration was not run when deploying `v258.  

TL;DR This index will not exist for anyone who's on a director that ever ran release `v257`

**Comment by tylerschultz** [comment]
@cnelson Good grief. Thanks for the detective work.

**Comment by tylerschultz** [comment]
Update: We're working on a fix to update this broken migration. The fix will be in release version 260.1. The fix to the migration will drop the index only if it exists.

Sorry to all for the inconvenience.

**Comment by cppforlife** [comment]
Fixed in 261.1 bosh release. @cnelson Thank you for helping out.
